[wip] specify layered grammar

dkochmanski opened the pull request.

dkochmanski committed:
documentation: reorganize sections and add doc-layered-grammer draft

An overview and a reference are both included from the main
document. Also lift core concept nodes in the documentation.

dkochmanski committed:
layered grammar: add incomplete implementation

Missing:
- documentation needs work
- protocol is not documented
- define-chart macro
- <geom> class is embrionic now
- <coords>, <scale> and <facet> are stubs
- no tests at all

dkochmanski committed:
squashme: layered-grammar: remove reader aest

It conflicts with the constructor function name.

dkochmanski committed:
squashme: layered-grammar: change collision-modifier prototype

Function must "know" the last value in order to properly adjust
position. Consider dodge or stack modifiers.

skempf commented:
I think you still need reader access for `aest` (and probably the other readers you removed as well). Locally I have changed the reader to `stat-aest` to test the statistical transformations mod and count. Do you want me to push the change, or do you have another idea in mind?

To be more explicit, when I call `(statistical-transformation *my-stat-mod* *my-data-frame*)`, the function `modify-row` needs to get the `aest` definition from the aest slot in the stat instance. You could use slot-value, too, but I normally use a reader and define my readers with the class in front, like `stat-aest` or `layer-aest`.  I added comments to the code below specifically.

skempf commented on Source/layered-grammar.lisp:
```diff
@@ -1,8 +1,108 @@
+   (indexes :initform nil)))
+
+(define-class <stat> ()
+  ((aest :initarg :aest)))
```
```  ((aest :initarg :aest :reader stat-aest)))```

dkochmanski replied:
as discussed in the issue thread (closing)

skempf commented on Source/layered-grammar.lisp:
```diff
@@ -1,8 +1,108 @@
+          (last-vals nil))
+      (flet ((modify-row (index row)
+               (declare (ignore index))
+               (let* ((aest (aest stat))
```
```               (let* ((aest (stat-aest stat))```

dkochmanski replied:
I've made function aest a dwim operator (as mentioned in the issue thread)

skempf commented on Source/statistical-transformations.lisp:
```diff
@@ -0,0 +1,26 @@
+    (prog1 df
+      (flet ((add-it (index row)
+               (declare (ignore index))
+               (let* ((vals (funcall aest df row))
```
```               (let* ((vals (map-aesthetics aest df row))```

dkochmanski replied:
fixed

skempf commented on Source/statistical-transformations.lisp:
```diff
@@ -0,0 +1,26 @@
+
+(defmethod statistical-transformation ((stat <count>) (frame <data-frame>))
+  (let ((df (copy-data-frame frame))
+        (aest (aest stat))
```
```        (aest (stat-aest stat))```

dkochmanski replied:
aest is a dwim operator now.

dkochmanski committed:
squashme: layered grammar: make aest a DWIM interface

When there is a single argument we assume it is an instance with a
slot 'aest and return it. Otherwise we create new aest object.

dkochmanski committed:
statistical-transformation: fix typo

we call map-aesthetics (not funcall).

dkochmanski committed:
layered-grammar: add common superclass and untie mods and stat

- <layered-grammar-comonent> is a superclass of stat, mods and
  geom (may be more later)

- <identity> is a constant which is dispatched in function stat and
  mdos to the correct class

- <mods> does not inherit from <stat>, if we need such feature we may
  readd it in the future (yagni)

dkochmanski committed:
layered grammar: remove unmap-aesthetics!

previous commit removed the method which used it, so we have no use
for this function now.

dkochmanski committed:
layered-grammar: add data to the <layered-grammar-component>

dkochmanski committed:
layered-grammar: aest: make operator a macro

dkochmanski committed:
layered-grammar: make stat, mods and geom macros

for consistency with aest and to cut off unnecessary apply.

skempf commented:
Do you intend the macro to be the function for the user? Is there a way to set the keys for specific classes? At the moment, I have something like this
```
(defun stat-bin (aest &key (width nil) (count 29))
  (<stat-bin> :aest aest :width width :count count))
```

dkochmanski commented:
I intend for macros to be operators used by the user, yes.

i.e
```
(defmacro stat (class-name &rest args)
  (case
    (<bin> (<stat-bin> ,@args))
    (otherwise (,class-name ,@args))))
```

modeline hints won't buy us much here because some aesthetics are required and some are optional, I'd ignore it for now. n.b allegro had a declaration for "full" lambda list.

ultimately we want to have a graphical tooling to plot data written in CLIM, then we may come up with pointer-documentation hints or other ways to direct the user at valid values.

dkochmanski committed:
layered-grammar: aest: add :enum qualifier for column name

For example

  (aest :x (:enum "implementation"))

It is similar to (factor "column") in ggplot2 - values are represented
as integers.

skempf committed:
data-frame: add copy-data-frame specialization on vector

skempf commented:
A few thoughts:
1. In the end, I have to copy the data anyway; is there much difference from consing and this? If not, this could be removed.
2. I could call ```<raw-data-frame>``` directly in the statistical-transformation instead, building a proper array. I wasn't sure if you wanted that much protocol separation or not.
3. I could also assume that the array being supplied is properly setup and just use it in call to ```<raw-data-frame>``` inside of copy-data-frame, i.e. using a displaced copy instead of using copy-array

I'll add documentation and testing, but wanted you to see this before I do that work.

skempf committed:
data-frame: remove copy-data-frame specialization on vector

This is not a great improvement, because the vectors still have to be
copied when making the data-frame.

skempf committed:
layered-grammar: add function to return copy of aest

Needed for 'bin statistical-transformation. Function also takes new
aesthetics to add to the new aest.

skempf committed:
layered-grammar: adding function to create empty data-frame from aest

skempf committed:
layered-grammar: Add function to collect the data ranges from data-frame using aest.

skempf committed:
statistical-transformation: adding bin transformation.

Two helper functions added to aid in getting bin centers for uniform and
non-uniform bins.

skempf commented:
I removed the copy-data-frame specialization, because I don't think it adds anything. I added all other changes separately. Adding:
1. copy-aest to return new copy that can be associated with the new df
2. make-empty-data-frame to return an empty data frame with the columns set
3. get-data-range to find the data ranges for all columns specified in aest
4. uniform-bins / non-uniform bins to get bin centers
5. the transformation itself.

skempf committed:
statistical-transformation: fixing bug whether left or right edge considered in bin.

Jitter (or nudge) factor should be multiplied by the median width.
Corrected this by adding second value to nonuniform-bins function to
return bin-widths.

skempf committed:
statistical-transformation: clean up comments.

skempf committed:
statistical-transformation: change bin and count names to be consistent with identity.

skempf committed:
layered-grammar: change bin and count names to be consistent with identity.

skempf committed:
statistical-transformation: modify stat-count to return new data-frame with only unique values.

The previous behavior was to return a copy of the incoming data-frame
and add a column to that with counts at each value of the :x aesthetic.
This produces a data-frame with (potentially) multiple :x values with
counts.

The new behaviour is to return a new data-frame with only unique values
of :x and the associated frequency of values at that location.
